### PR TITLE
ENG-2654 ENG-2675 ENG-2651 ENG-2683 ENG-2653 Implement Hub registry features into ECR screen

### PR DESCRIPTION
**Pull request by ichalagashvili**
This PR contains work for:
https://entando.myjetbrains.com/youtrack/issue/ENG-2654
https://entando.myjetbrains.com/youtrack/issue/ENG-2675
https://entando.myjetbrains.com/youtrack/issue/ENG-2651
https://entando.myjetbrains.com/youtrack/issue/ENG-2683
https://entando.myjetbrains.com/youtrack/issue/ENG-2653

### CODE REVIEW

**Comment by entando-jx** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To fully approve this pull request, please assign additional approvers.
We suggest the following additional approver: **ichalagashvili**

If they are not already assigned, you can assign the PR to them by writing `/assign @ichalagashvili` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=entando%2Fapp-builder).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/entando/app-builder/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: sass/component-repository/components/HubRegistrySwitcher.scss
Comment (jeffgo10) [inline sass/component-repository/components/HubRegistrySwitcher.scss:111]: add newline

**Review by jeffgo10 — DISMISSED** [review]
looks good. only this one simple change

**Comment by entando-jx** [comment]
@ichalagashvili: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
pr-build | 964101cf7a0cad4a08d943cda226ff2d250c8192 | [link]() | `/test this`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Review by raxkaynan — CHANGES_REQUESTED** [review]
There seem to be no tests.

File: src/state/component-repository/hub/actions.js
Comment (ichalagashvili) [inline src/state/component-repository/hub/actions.js:273]: ignore this. it will be added later (BE not yet implemented)

File: src/api/component-repository/hub.js
Comment (ichalagashvili) [inline src/api/component-repository/hub.js:86]: ignore this please. BE not yet implemented

File: config/webpack.config.dev.js
Comment (raxkaynan) [inline config/webpack.config.dev.js]: These could may be better below the default `publicPath` definition since the comment interposes between the Webpack comment and the default definition, which could cause confusion.

File: config/webpack.config.dev.js
Comment (raxkaynan) [inline config/webpack.config.dev.js]: Same as the above comment. But this even seems more confusing since before the ecr/hub comment, it says "Omit trailing slash...".

File: sass/component-repository/components/HubRegistrySwitcher.scss
Comment (raxkaynan) [inline sass/component-repository/components/HubRegistrySwitcher.scss]: This could be just `//sass-lint:disable-line...`

File: sass/index.scss
Comment (raxkaynan) [inline sass/index.scss]: For consistency, we should just exclude the `.scss` here and in the lines below.

File: src/api/component-repository/components.js
Comment (raxkaynan) [inline src/api/component-repository/components.js:16]: Is `repoUrl` part of the path (i.e. a path variable) or should it be a query parameter?

File: src/app-init/router.js
Comment (raxkaynan) [inline src/app-init/router.js]: Why is a trailing slash added here?

File: src/locales/it.js
Comment (raxkaynan) [inline src/locales/it.js]: Should we not translate this (and others below) to Italian?

File: src/locales/pt.js
Comment (raxkaynan) [inline src/locales/pt.js]: Should we not translate this (and others below) to Portuguese?

File: src/state/component-repository/hub/actions.js
Comment (raxkaynan) [inline src/state/component-repository/hub/actions.js:202]: It's probably better to avoid calling a "fetch" action in a delete action and instead dispatch an action that updates the state (i.e. a `removeXX()` action the same as others).

File: src/state/component-repository/hub/actions.js
Comment (raxkaynan) [inline src/state/component-repository/hub/actions.js]: The naming is quite inconsistent with others. This could be changed to `sendPostRegistry` or simply just `addRegistry`.

File: src/state/component-repository/hub/actions.js
Comment (raxkaynan) [inline src/state/component-repository/hub/actions.js:223]: Similar case to the delete action above. It should be better to update the state instead of calling a fetch action here.

File: src/ui/component-repository/components/list/AddNewRegistryModal.js
Comment (raxkaynan) [inline src/ui/component-repository/components/list/AddNewRegistryModal.js]: The "is" in `isUnique` implies that this returns a boolean value. This seems like a validation function and in that case the conventional way we name it is just what the validation rule describes, such as `required`, in order to distinguish it from a boolean expression/function.

File: src/ui/component-repository/components/list/BundleListGridView.js
Comment (raxkaynan) [inline src/ui/component-repository/components/list/BundleListGridView.js]: `i % 2 === 0` means that it's even, not odd :)

File: src/ui/component-repository/components/list/BundleListGridView.js
Comment (raxkaynan) [inline src/ui/component-repository/components/list/BundleListGridView.js]: `tabIndex="-1"` already suffices. No need to multiply it by `i`.

File: src/ui/component-repository/components/list/BundleListListView.js
Comment (raxkaynan) [inline src/ui/component-repository/components/list/BundleListListView.js]: `tabIndex="-1"` already suffices. No need to multiply it by `i`.

File: src/ui/component-repository/components/list/ComponentListGridView.js
Comment (raxkaynan) [inline src/ui/component-repository/components/list/ComponentListGridView.js]: `tabIndex="-1"` already suffices. No need to multiply it by `i`.

File: src/ui/component-repository/components/list/ComponentListGridView.js
Comment (raxkaynan) [inline src/ui/component-repository/components/list/ComponentListGridView.js]: `tabIndex="-1"` already suffices. No need to multiply it by `i`.

File: src/ui/component-repository/components/list/ComponentListListView.js
Comment (raxkaynan) [inline src/ui/component-repository/components/list/ComponentListListView.js]: `tabIndex="-1"` already suffices. No need to multiply it by `i`.

File: src/ui/component-repository/components/list/ComponentListListView.js
Comment (raxkaynan) [inline src/ui/component-repository/components/list/ComponentListListView.js]: `tabIndex="-1"` already suffices. No need to multiply it by `i`.

File: src/ui/component-repository/components/list/HubRegistrySwitcher.js
Comment (raxkaynan) [inline src/ui/component-repository/components/list/HubRegistrySwitcher.js]: `tabIndex="-1"` is the common practice to remove the element from the default keyboard navigation flow.

File: src/ui/component-repository/components/list/HubRegistrySwitcher.js
Comment (raxkaynan) [inline src/ui/component-repository/components/list/HubRegistrySwitcher.js:88]: `tabIndex="-1"` already suffices. No need to multiply it by `i`.

File: src/ui/component-repository/components/list/BundleListGridView.js
Comment (raxkaynan) [inline src/ui/component-repository/components/list/BundleListGridView.js]: Fragment doesn't seem necessary here.

File: src/ui/component-repository/components/list/BundleListGridView.js
Comment (raxkaynan) [inline src/ui/component-repository/components/list/BundleListGridView.js:64]: Is this always not displayed?

File: src/ui/component-repository/components/list/BundleListListView.js
Comment (raxkaynan) [inline src/ui/component-repository/components/list/BundleListListView.js]: Fragment doesn't seem to be necessary here.

File: src/ui/component-repository/components/list/BundleListListView.js
Comment (raxkaynan) [inline src/ui/component-repository/components/list/BundleListListView.js:55]: Is this always not displayed?

File: src/ui/component-repository/components/list/ComponentListGridView.js
Comment (raxkaynan) [inline src/ui/component-repository/components/list/ComponentListGridView.js]: Same as previous comment. `i % 2 === 0` checks if `i` is even, not odd.

File: src/ui/component-repository/components/list/ComponentListGridView.js
Comment (raxkaynan) [inline src/ui/component-repository/components/list/ComponentListGridView.js:82]: Is this always not displayed?

File: src/ui/component-repository/components/list/ComponentListListView.js
Comment (raxkaynan) [inline src/ui/component-repository/components/list/ComponentListListView.js:74]: Is this always not displayed?

File: src/ui/component-repository/components/list/HubBundleList.js
Comment (raxkaynan) [inline src/ui/component-repository/components/list/HubBundleList.js]: `useMemo()` is unnecessary here since this is merely assignment of an object.

File: src/ui/component-repository/components/list/HubRegistrySwitcher.js
Comment (raxkaynan) [inline src/ui/component-repository/components/list/HubRegistrySwitcher.js:23]: Not sure why you're using `useCallback()` here and not with the other handlers below.

File: src/ui/component-repository/components/list/HubRegistrySwitcher.js
Comment (raxkaynan) [inline src/ui/component-repository/components/list/HubRegistrySwitcher.js:46]: Why should exhaustive-deps rule be disabled here? Is is strongly recommended to keep it enabled.

**Review by raxkaynan — CHANGES_REQUESTED** [review]

File: config/webpack.config.dev.js
Comment (ichalagashvili) [reply config/webpack.config.dev.js]: agree. done ✅

File: config/webpack.config.dev.js
Comment (ichalagashvili) [reply config/webpack.config.dev.js]: agree done ✅

File: sass/component-repository/components/HubRegistrySwitcher.scss
Comment (ichalagashvili) [reply sass/component-repository/components/HubRegistrySwitcher.scss]: agree. done ✅

File: sass/index.scss
Comment (ichalagashvili) [reply sass/index.scss]: agree. done ✅

File: src/api/component-repository/components.js
Comment (ichalagashvili) [reply src/api/component-repository/components.js:16]: `repoUrl=` is part of the path. Thats according to a new BE API approach.

File: src/app-init/router.js
Comment (ichalagashvili) [reply src/app-init/router.js]: agree. removed ✅

File: src/locales/it.js
Comment (ichalagashvili) [reply src/locales/it.js]: agree done ✅

File: src/locales/pt.js
Comment (ichalagashvili) [reply src/locales/pt.js]: agree done ✅

File: src/state/component-repository/hub/actions.js
Comment (ichalagashvili) [reply src/state/component-repository/hub/actions.js:202]: Well I think its common method to re-fetch stuff after deletion. Also, if you look at the codebase at many places we refetch stuff. For example when adding/deleting a category:
```
export const sendPostCategory = categoryData => (dispatch, getState) => {
  try {
    const { parentCode } = categoryData;
    return dispatch(wrapApiCall(postCategory)(categoryData)).then((data) => {
      dispatch(setCategories([data.payload]));
      const { parentCode: grandParentCode } = getCategoriesMap(getState())[parentCode];
      dispatch(fetchCategoryTree(grandParentCode));
      history.push(ROUTE_CATEGORY_LIST);
    });
  } catch (e) {
    return Promise.resolve();
  }
};
```
Its much up-to-date that way and avoid any edge cases. Please dont be mad if I keep it 🤚

File: src/state/component-repository/hub/actions.js
Comment (ichalagashvili) [reply src/state/component-repository/hub/actions.js]: agree done ✅

File: src/state/component-repository/hub/actions.js
Comment (ichalagashvili) [reply src/state/component-repository/hub/actions.js:223]: I am keen to keep this because also for example when a registry is added, newly added registry has some additional data (for example id) that must be fetched.

File: src/ui/component-repository/components/list/AddNewRegistryModal.js
Comment (ichalagashvili) [reply src/ui/component-repository/components/list/AddNewRegistryModal.js]: agree , renamed to `mustBeUnique`

File: src/ui/component-repository/components/list/BundleListGridView.js
Comment (ichalagashvili) [reply src/ui/component-repository/components/list/BundleListGridView.js]: agree, done ✅

File: src/ui/component-repository/components/list/BundleListGridView.js
Comment (ichalagashvili) [reply src/ui/component-repository/components/list/BundleListGridView.js]: done ✅

File: src/ui/component-repository/components/list/BundleListListView.js
Comment (ichalagashvili) [reply src/ui/component-repository/components/list/BundleListListView.js]: done ✅

File: src/ui/component-repository/components/list/ComponentListGridView.js
Comment (ichalagashvili) [reply src/ui/component-repository/components/list/ComponentListGridView.js]: done ✅

File: src/ui/component-repository/components/list/ComponentListGridView.js
Comment (ichalagashvili) [reply src/ui/component-repository/components/list/ComponentListGridView.js]: done ✅

File: src/ui/component-repository/components/list/ComponentListListView.js
Comment (ichalagashvili) [reply src/ui/component-repository/components/list/ComponentListListView.js]: done ✅

File: src/ui/component-repository/components/list/ComponentListListView.js
Comment (ichalagashvili) [reply src/ui/component-repository/components/list/ComponentListListView.js]: done ✅

File: src/ui/component-repository/components/list/HubRegistrySwitcher.js
Comment (ichalagashvili) [reply src/ui/component-repository/components/list/HubRegistrySwitcher.js]: done ✅

File: src/ui/component-repository/components/list/BundleListListView.js
Comment (ichalagashvili) [reply src/ui/component-repository/components/list/BundleListListView.js:55]: yes. We needed to have a container live outside modal so a new modal did not die once parent dies. Its a bit of hack we had to come up with. But its 1000% intended.

File: src/ui/component-repository/components/list/BundleListGridView.js
Comment (ichalagashvili) [reply src/ui/component-repository/components/list/BundleListGridView.js:64]: yes. We needed to have a container live outside modal so a new modal did not die once parent dies. Its a bit of hack we had to come up with. But its 1000% intended.

File: src/ui/component-repository/components/list/BundleListGridView.js
Comment (ichalagashvili) [reply src/ui/component-repository/components/list/BundleListGridView.js]: done ✅

File: src/ui/component-repository/components/list/BundleListListView.js
Comment (ichalagashvili) [reply src/ui/component-repository/components/list/BundleListListView.js]: done ✅

File: src/ui/component-repository/components/list/ComponentListGridView.js
Comment (ichalagashvili) [reply src/ui/component-repository/components/list/ComponentListGridView.js:82]: yes. We needed to have a container live outside modal so a new modal did not die once parent dies. Its a bit of hack we had to come up with. But its 1000% intended.

File: src/ui/component-repository/components/list/ComponentListListView.js
Comment (ichalagashvili) [reply src/ui/component-repository/components/list/ComponentListListView.js:74]: yes. We needed to have a container live outside modal so a new modal did not die once parent dies. Its a bit of hack we had to come up with. But its 1000% intended.

File: src/ui/component-repository/components/list/ComponentListGridView.js
Comment (ichalagashvili) [reply src/ui/component-repository/components/list/ComponentListGridView.js]: done ✅

File: src/ui/component-repository/components/list/HubBundleList.js
Comment (ichalagashvili) [reply src/ui/component-repository/components/list/HubBundleList.js]: agree done ✅

File: src/ui/component-repository/components/list/HubRegistrySwitcher.js
Comment (ichalagashvili) [reply src/ui/component-repository/components/list/HubRegistrySwitcher.js:23]: done. Wrapped others in `useCallback` too ✅

File: src/ui/component-repository/components/list/HubRegistrySwitcher.js
Comment (ichalagashvili) [reply src/ui/component-repository/components/list/HubRegistrySwitcher.js:46]: removed, done ✅

Comment (ichalagashvili) [reply src/ui/component-repository/components/list/HubRegistrySwitcher.js:46]: I cannot remove this. As I need that effect to run *only* when mounting (empty array)

Comment (ichalagashvili) [reply src/ui/component-repository/components/list/HubRegistrySwitcher.js:46]: When I am removing the eslint line, it automatically populated with deps array... which I dont want. I want explicitly when a component mounts, no more.

Comment (raxkaynan) [reply src/ui/component-repository/components/list/HubRegistrySwitcher.js:46]: It will indeed just run once when you put the required dependencies unless those dependencies change. In that case, it is expected since it makes sure that the component doesn't show stale data. It is strongly discouraged by the React team to bypass that rule. However, if you really are sure that only the initial values of the props/state used in the "effect" are needed and that the function should strictly only run once, then I believe we can let this pass for now.

File: src/state/component-repository/hub/actions.js
Comment (raxkaynan) [reply src/state/component-repository/hub/actions.js:202]: The `sendDeleteCategory` doesn't fetch after deletion. But yeah in some actions it does. Although I'd consider fetching after adding/posting data is more acceptable since there are cases when the API response doesn't include the newly created data. Anyway, one key concern here is only on performance (both in prod & e2e tests), since you're fetching everything again even though just one part of the data has been modified/added/deleted. But since I don't think that it significantly impairs performance (at least for now), then it's fine with me.

**Review by raxkaynan — APPROVED** [review]
Looks good. Just some general feedback but approved.

**Pull request merged**
